File: _skels/go-ddd-skel/internal/catalog/service.go
// Service-layer logic for /api/catalog.
package catalog

import (
	"context"
	"fmt"

	"github.com/example/go-ddd-skel/internal/shared"
)

// Service coordinates catalog.Repository for HTTP routes + the
// cross-resource orders flow (which holds a Repository directly).
type Service struct {
	repo Repository
}

// NewService builds a catalog Service.
func NewService(repo Repository) *Service {
	return &Service{repo: repo}
}

// List returns every catalog item.
func (s *Service) List(ctx context.Context) ([]CatalogItem, error) {
	return s.repo.List(ctx)
}

// Get fetches a catalog item by id.
func (s *Service) Get(ctx context.Context, id uint) (CatalogItem, error) {
	return s.repo.Get(ctx, id)
}

// Create inserts a new catalog item.
func (s *Service) Create(ctx context.Context, dto CreateCatalogItemDTO) (CatalogItem, error) {
	if dto.Name == "" {
		return CatalogItem{}, fmt.Errorf("%w: catalog item name cannot be empty", shared.ErrValidation)
	}
	item := CatalogItem{
		Name:      dto.Name,
		Price:     dto.Price,
		Available: true,
	}
	if dto.Description != nil {
		item.Description = *dto.Description
	}
	if dto.Category != nil {
		item.Category = *dto.Category
	}
	if dto.Available != nil {
		item.Available = *dto.Available
	}
	if err := s.repo.Save(ctx, &item); err != nil {
		return CatalogItem{}, err
	}
	return item, nil
}

File: _skels/rust-actix-ddd-skel/src/catalog/service.rs
//! Service-layer logic for `/api/catalog`.

use std::sync::Arc;

use crate::catalog::repository::{CatalogItem, CatalogRepository, NewCatalogItem};
use crate::shared::DomainError;

#[derive(Debug, Clone)]
pub struct NewCatalogItemDTO {
    pub name: String,
    pub description: Option<String>,
    pub price: f64,
    pub category: Option<String>,
    pub available: bool,
}

#[derive(Clone)]
pub struct CatalogService {
    repo: Arc<dyn CatalogRepository>,
}

impl CatalogService {
    pub fn new(repo: Arc<dyn CatalogRepository>) -> Self {
        Self { repo }
    }

    pub async fn list(&self) -> Result<Vec<CatalogItem>, DomainError> {
        self.repo.list().await
    }

    pub async fn get(&self, id: i32) -> Result<CatalogItem, DomainError> {
        self.repo.get(id).await
    }

    pub async fn create(&self, dto: NewCatalogItemDTO) -> Result<CatalogItem, DomainError> {
        if dto.name.trim().is_empty() {
            return Err(DomainError::Validation(
                "catalog item name cannot be empty".to_string(),
            ));
        }
        if dto.price < 0.0 {
            return Err(DomainError::Validation(
                "price cannot be negative".to_string(),
            ));
        }
        self.repo
            .create(NewCatalogItem {
                name: dto.name,
                description: dto.description.unwrap_or_default(),
                price: dto.price,
                category: dto.category.unwrap_or_default(),
                available: dto.available,
            })
            .await
    }
}
